File: cpp/expert_trading_features/01_fixed_point_price.cpp
// ============================================================================
// 01 — Fixed-point prices: never represent money/prices with floating point
// Compile:  clang++ -std=c++20 -O2 01_fixed_point_price.cpp -o bin/01 && bin/01
// ----------------------------------------------------------------------------
// Why this matters in trading:
//   * `double` cannot represent 0.10 exactly -> rounding drift, failed equality,
//     and prices that don't land on the exchange tick grid.
//   * Exchanges quote prices on an integer *tick* grid (e.g. $0.01). Store the
//     price AS an integer number of ticks (or scaled integer) and all arithmetic
//     becomes exact, branch-free, and trivially comparable.
//   * Integer compares/adds are also faster and have deterministic latency.
// ============================================================================

#include <cstdint>
#include <iostream>
#include <format>
#include <compare>
#include <chrono>

// A price scaled to 4 implied decimals (1 unit = 0.0001). 123.4500 -> 1234500.
// int64 holds ~±9.2e14 at this scale = ±922,337,203,685 dollars. Plenty.
class Price {
    std::int64_t ticks_ = 0;                 // scaled integer, the single source of truth
    static constexpr std::int64_t kScale = 10'000;   // 4 decimal places

public:
    Price() = default;
    constexpr explicit Price(std::int64_t scaled) : ticks_(scaled) {}

    // Build from a human "dollars and 1/10000ths" without ever touching double.
    static constexpr Price from_decimal(std::int64_t whole, std::int64_t frac4 = 0) {
        return Price(whole * kScale + frac4);
    }

    constexpr std::int64_t raw() const { return ticks_; }

    // Exact arithmetic — these are just integer ops.
    constexpr Price operator+(Price o) const { return Price(ticks_ + o.ticks_); }
    constexpr Price operator-(Price o) const { return Price(ticks_ - o.ticks_); }

    // Spaceship => exact, total ordering for free (sorting a book, equality checks).
    constexpr auto operator<=>(const Price&) const = default;

    // Snap to a coarser tick size (e.g. round a theo price to the $0.01 grid).
    constexpr Price round_to_tick(std::int64_t tick_in_scale) const {
        // round-half-up on the integer grid, no FP.
        std::int64_t half = tick_in_scale / 2;
        std::int64_t r = ((ticks_ + (ticks_ >= 0 ? half : -half)) / tick_in_scale) * tick_in_scale;
        return Price(r);
    }

    // Only convert to text at the very edge (logging / display).
    std::string to_string() const {
        std::int64_t w = ticks_ / kScale, f = std::llabs(ticks_ % kScale);
        return std::format("{}.{:04}", w, f);
    }
};

int main() {
    // --- The classic floating-point trap --------------------------------------
    double a = 0.1, b = 0.2;
    std::cout << std::format("double 0.1 + 0.2 == 0.3 ? {}   (actual = {:.17f})\n",
                             (a + b == 0.3), a + b);

    // --- Fixed-point is exact -------------------------------------------------
    Price p1 = Price::from_decimal(0, 1000);   // 0.1000
    Price p2 = Price::from_decimal(0, 2000);   // 0.2000
    Price p3 = Price::from_decimal(0, 3000);   // 0.3000
    std::cout << std::format("fixed  0.1 + 0.2 == 0.3 ? {}   (sum = {})\n",
                             (p1 + p2) == p3, (p1 + p2).to_string());

    // --- Spread + tick rounding ----------------------------------------------
    Price bid = Price::from_decimal(123, 4500);   // 123.4500
    Price ask = Price::from_decimal(123, 4600);   // 123.4600
    Price spread = ask - bid;
    std::cout << std::format("bid={} ask={} spread={}\n",
                             bid.to_string(), ask.to_string(), spread.to_string());

    Price theo = Price::from_decimal(123, 4537);                 // 123.4537
    Price penny = theo.round_to_tick(100);                       // round to $0.01
    std::cout << std::format("theo={} rounded-to-$0.01={}\n",
                             theo.to_string(), penny.to_string());

    // --- Compile-time pricing (zero runtime cost) -----------------------------
    constexpr Price limit = Price::from_decimal(100, 0) + Price::from_decimal(0, 2500);
    static_assert(limit.raw() == 1'002'500, "computed entirely at compile time");
    std::cout << std::format("constexpr limit price = {}\n", limit.to_string());

    // --- Micro-benchmark: integer-tick vs double arithmetic -------------------
    // The headline reason for fixed-point is CORRECTNESS, not speed. But the
    // integer path is also branch-free and deterministic, so we should give up
    // nothing on throughput either. We time a tight add + compare loop both ways.
    // (At -O2 this is often a near-tie — the point is exactness comes for free.)
    constexpr std::size_t N = 100'000'000;

    Price acc_fp{0};
    const Price step_fp = Price::from_decimal(0, 1);    // +0.0001 per iter
    std::int64_t hits_fp = 0;
    auto t0 = std::chrono::steady_clock::now();
    for (std::size_t i = 0; i < N; ++i) {
        acc_fp = acc_fp + step_fp;
        if (acc_fp >= step_fp) ++hits_fp;               // exact integer compare
    }
    auto t1 = std::chrono::steady_clock::now();

    double acc_d = 0.0;
    const double step_d = 0.0001;
    std::int64_t hits_d = 0;
    for (std::size_t i = 0; i < N; ++i) {
        acc_d += step_d;
        if (acc_d >= step_d) ++hits_d;                  // FP compare (rounding!)
    }
    auto t2 = std::chrono::steady_clock::now();

    double ns_fp = std::chrono::duration<double, std::nano>(t1 - t0).count() / N;
    double ns_d  = std::chrono::duration<double, std::nano>(t2 - t1).count() / N;
    std::cout << std::format(
        "\nbench add+compare over {} iters:\n"
        "  fixed-point : {:.3f} ns/op  (acc={}, hits={})\n"
        "  double      : {:.3f} ns/op  (acc={:.4f}, hits={})\n"
        "  ratio       : {:.2f}x (fixed-point gives exactness at ~no speed cost)\n",
        N, ns_fp, acc_fp.to_string(), hits_fp, ns_d, acc_d, hits_d, ns_d / ns_fp);

    return 0;
}

File: rust/expert_trading_features/examples/01_fixed_point_price.rs
// ============================================================================
// 01 — Fixed-point prices: never represent money/prices with floating point
// Run:  cargo run --release --example 01_fixed_point_price
// C++ counterpart: cpp/expert_trading_features/01_fixed_point_price.cpp
// ----------------------------------------------------------------------------
// Why this matters in trading:
//   * `f64` cannot represent 0.10 exactly -> rounding drift, failed equality, and
//     prices that don't land on the exchange tick grid.
//   * Exchanges quote on an integer *tick* grid. Store the price AS a scaled
//     integer and all arithmetic is exact, branch-free, and trivially comparable.
//
// What Rust changes vs C++:
//   * A `#[repr(transparent)]` newtype makes `Px` a DISTINCT type from a bare i64
//     (and from `Qty`), so the compiler rejects mixing a price with a quantity —
//     a whole class of bugs gone, at zero runtime cost (same layout as i64).
//   * `derive(PartialOrd, Ord, Eq)` gives exact, total ordering for free.
//   * Arithmetic that can overflow is opt-in explicit (`checked_`/`saturating_`),
//     and debug builds panic on overflow instead of silently wrapping.
// ============================================================================

use std::fmt;

/// A price scaled to 4 implied decimals (1 unit = 0.0001). 123.4500 -> 1_234_500.
/// `repr(transparent)` => identical layout/ABI to `i64`; the newtype is free.
#[repr(transparent)]
#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
struct Px(i64);

/// A separate unit for size. Because `Px` and `Qty` are different types, you
/// cannot accidentally pass a quantity where a price is expected.
#[repr(transparent)]
#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Default)]
struct Qty(u64);

impl Px {
    const SCALE: i64 = 10_000; // 4 decimal places

    const fn from_decimal(whole: i64, frac4: i64) -> Px {
        Px(whole * Px::SCALE + frac4)
    }
    const fn raw(self) -> i64 {
        self.0
    }

    // Exact arithmetic. `checked_add` returns None on overflow rather than wrapping
    // or invoking UB — you decide the policy at the call site.
    fn checked_add(self, rhs: Px) -> Option<Px> {
        self.0.checked_add(rhs.0).map(Px)
    }

    // Snap to a coarser tick (e.g. round a theo to the $0.01 grid), round-half-up,
    // entirely on the integer grid — no float ever touches a price.
    fn round_to_tick(self, tick: i64) -> Px {
        let half = tick / 2;
        let bias = if self.0 >= 0 { half } else { -half };
        Px(((self.0 + bias) / tick) * tick)
    }
}

// Operator overloading is just a trait. We only implement the ops that make sense
// for a price (Px + Px, Px - Px) — you can't multiply two prices together.
impl std::ops::Add for Px {
    type Output = Px;
    fn add(self, o: Px) -> Px {
        Px(self.0 + o.0)
    }
}
impl std::ops::Sub for Px {
    type Output = Px;
    fn sub(self, o: Px) -> Px {
        Px(self.0 - o.0)
    }
}

// Only convert to text at the edge (logging / display).
impl fmt::Display for Px {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let w = self.0 / Px::SCALE;
        let frac = (self.0 % Px::SCALE).abs();
        write!(f, "{w}.{frac:04}")
    }
}
impl fmt::Debug for Px {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Px({self})")
    }
}

fn main() {
    // The classic floating-point trap.
    let a = 0.1f64 + 0.2;
    println!("f64   0.1 + 0.2 == 0.3 ? {}   (actual = {a:.17})", a == 0.3);

    // Fixed-point is exact.
    let p1 = Px::from_decimal(0, 1000); // 0.1000
    let p2 = Px::from_decimal(0, 2000); // 0.2000
    let p3 = Px::from_decimal(0, 3000); // 0.3000
    println!(
        "fixed 0.1 + 0.2 == 0.3 ? {}   (sum = {})",
        p1 + p2 == p3,
        p1 + p2
    );

    let bid = Px::from_decimal(123, 4500);
    let ask = Px::from_decimal(123, 4600);
    println!("bid={bid} ask={ask} spread={}", ask - bid);

    let theo = Px::from_decimal(123, 4537);
    println!("theo={theo} rounded-to-$0.01={}", theo.round_to_tick(100));

    // The unit safety: this line does not compile — Px and Qty are distinct types.
    //   let _ = bid + Qty(100);
    let size = Qty(100);
    println!(
        "(price {bid} and qty {} are different types — can't be mixed)",
        size.0
    );

    // `const fn` => compile-time pricing, zero runtime cost.
    const LIMIT: Px = Px::from_decimal(100, 2500);
    const _: () = assert!(LIMIT.raw() == 1_002_500);
    println!("const limit price = {LIMIT}");

    // Overflow is explicit: checked_add returns None instead of wrapping/UB.
    println!(
        "checked_add near MAX: {:?}",
        Px(i64::MAX).checked_add(Px(1))
    );
    println!(
        "checked_add normal:   {:?}",
        bid.checked_add(Px::from_decimal(0, 100))
    );
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn addition_is_exact() {
        assert_eq!(
            Px::from_decimal(0, 1000) + Px::from_decimal(0, 2000),
            Px::from_decimal(0, 3000)
        );
    }

    #[test]
    fn ordering_is_total_and_exact() {
        let mut v = [
            Px::from_decimal(1, 0),
            Px::from_decimal(0, 9999),
            Px::from_decimal(1, 1),
        ];
        v.sort();
        assert_eq!(
            v,
            [
                Px::from_decimal(0, 9999),
                Px::from_decimal(1, 0),
                Px::from_decimal(1, 1)
            ]
        );
    }

    #[test]
    fn round_to_penny() {
        assert_eq!(
            Px::from_decimal(123, 4537).round_to_tick(100),
            Px::from_decimal(123, 4500)
        );
        assert_eq!(
            Px::from_decimal(123, 4567).round_to_tick(100),
            Px::from_decimal(123, 4600)
        );
    }

    #[test]
    fn checked_add_flags_overflow() {
        assert_eq!(Px(i64::MAX).checked_add(Px(1)), None);
        assert_eq!(Px(1).checked_add(Px(2)), Some(Px(3)));
    }

    #[test]
    fn newtype_is_zero_cost() {
        assert_eq!(std::mem::size_of::<Px>(), std::mem::size_of::<i64>());
    }
}
